misk-action-scopes/src/main/kotlin/misk/scope/ActionScope.kt

```diff
@@ -20,37 +21,58 @@ class ActionScope @Inject internal constructor(
   private val providers: @JvmSuppressWildcards Map<Key<*>, Provider<ActionScopedProvider<*>>>
 ) : AutoCloseable {
   companion object {
-    private val tls = ThreadLocal<LinkedHashMap<Key<*>, Any?>>()
+    private val threadLocalScope = ThreadLocal<LinkedHashMap<Key<*>, Any?>>()
+    private val threadLocalUUID = ThreadLocal<UUID>()
   }
 
   /** Starts the scope on a thread with the provided seed data */
   fun enter(seedData: Map<Key<*>, Any?>): ActionScope {
-    check(tls.get() == null) {
+    check(!inScope()) {
       "cannot begin an ActionScope on a thread that is already running in an action scope"
     }
 
-    tls.set(LinkedHashMap(seedData))
+    threadLocalScope.set(LinkedHashMap(seedData))
+
+    // If an action scope had previously been entered on the thread, re-use its UUID.
+    // Otherwise, generate a new one.
+    if (threadLocalUUID.get() == null) {
+      threadLocalUUID.set(UUID.randomUUID())
+    }
+
     return this
   }
 
   override fun close() {
-    tls.remove()
+    threadLocalScope.remove()
+
+    // Explicitly NOT removing threadLocalUUID because we want to retain the thread's UUID if
+    // the action scope is re-entered on the same thread.
+    // The only way in which threadLocalUUID is removed is through garbage collection, which occurs
+    // when the thread is no longer alive.
   }
 
   /** Returns true if currently in the scope */
-  fun inScope(): Boolean = tls.get() != null
+  fun inScope(): Boolean = threadLocalScope.get() != null
 
   /**
    * Wraps a [Callable] that will be called on another thread, propagating the current
    * scoped data onto that thread
    */
   fun <T> propagate(c: Callable<T>): Callable<T> {
-    check(tls.get() != null) { "not running within an ActionScope" }
+    check(inScope()) { "not running within an ActionScope" }
+
+    val currentScopedData = threadLocalScope.get().toMap()
+    val currentThreadUUID = threadLocalUUID.get()
 
-    val currentScopedData = tls.get().toMap()
     return Callable {
-      enter(currentScopedData).use {
-        c.call()
+      // If the original thread is the same as the thread that calls the Callable and we are already
+      // in scope, then there is no need to re-enter the scope.
+      if (inScope() && currentThreadUUID == threadLocalUUID.get()) {
+        return@Callable c.call()
```

Comment:
Forgot to clean this up while I was debugging. It should be:
```kotlin
if (inScope() && currentThreadUUID == threadLocalUUID.get()) {
  c.call()
} else {
  enter(currentScopedData).use {
    c.call()
  }
}
```

I'll add an additional commit.